docs/knowledgebase/runtime/benchmarking.md

```diff
@@ -0,0 +1,547 @@
+referenced hardware.
+
+So we want to have an estimate of how much computation it takes before actually running the
+extrinsics, and this will also affect how much transaction fee we charge beforehand. If the
+extrinsic turns out to run with less computation, some of the estimated weight can be returned, and
+eventually causing the transaction fee returned back to user. This is further explained in the
+[Transaction Fees chapter](./fees).
+
+Now the question is how do we determine the computation time and weight of our extrinsics? There
+might also be a relationship between the time and the input or the existing state (the current
+database storage) when running the extrinsic. As any computer scientist can tell, we are getting
+into a time complexity analysis of the function.
+
+This is where Substrate Runtime Benchmarking comes in. It has a set of toolchain to help determine
+the computation time and thus the transaction weight of a runtime extrinsic. It runs the benchmarked
+extrinsic over and over with varying inputs and the computation time is measured. At the end, it:
+
+- returns the raw data of these measured time with these varying input and how many database read
+  write has been performed.
+
+- deduces the coefficient on how the computation time changes with respect to the input using linear
+  regression, assuming a linear relationship. It is up to the runtime developer to review the raw
+  data and check if it is indeed a linear relationship. If not, either optimize the extrinsic for
+  its time complexity or adjust the weight function accordingly.
+
+- outputs a rust file with these weight functions that can be easily integrated in our runtime.
+
+## Why Benchmark a Runtime Pallet?
+
+Denial-of-service (DoS) is a common attack vector for distributed systems, including blockchain
+networks. A simple example would be for any user repeatedly requesting a function that involve
+intensive computation. An effective counter-measure is that such a request would bear a certain cost
+to the user. For a fair system, the user cost should reflect the computation and storage cost
+incurred to the system. To encourage users to use such system, we want to estimate this cost
+relatively accurately and not overcharging users.
+
+Now with Substrate benchmarking toolchain, runtime developers can estimate the weight of their
+runtime extrinsics which translate to the transaction fee charged to the runtime users. So it is
+critical to benchmark our runtime extrinsics to measure how the extrinsic computation changes with
+repect to its input and set the proper weight functions for these extrinsics. Setting a proper
+weight function that reflects accurately on the underlying computation and storage is an important
+security safeguard in Substrate.
+
+## Features of Substrate Benchamrking Toolchain
+
+- You can configure multiple input variables to be used when benchmarking an extrinsic. For each
```

Comment:
would be good to link to the `benchmarks!` macro do somewhere here, it as very good documentation.

Reply:
added on L#85